File: src/714. Best Time to Buy and Sell Stock with Transaction Fee.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */

export function maxProfit(prices, fee) {
  const n = prices.length
  let prevSell = 0,
    prevBuy = -prices[0]
  for (let i = 1; i < n; i++) {
    let currentSell = Math.max(prevSell, prevBuy + prices[i] - fee),
      currentBuy = Math.max(prevBuy, prevSell - prices[i])
    prevSell = currentSell
    prevBuy = currentBuy
  }
  return prevSell
}

File: src/714. Best Time to Buy and Sell Stock with Transaction Fee.cpp
class Solution {
    public:
        int maxProfit(vector < int > & prices, int fee) {
            int n = prices.size();
            int prevSell = 0;
            int prevBuy = -prices[0];
            for (int i = 0; i < n; ++i) {
                int currentSell = max(prevSell, prevBuy + prices[i] - fee);
                int currentBuy = max(prevBuy, prevSell - prices[i]);
                prevSell = currentSell;
                prevBuy = currentBuy;
            }
            return prevSell;
        }
};